libs/xpdf/xpdf-src/doc/pdftotext.1
.\" Copyright 1997-2025 Glyph & Cog, LLC
.TH pdftotext 1 "14 Nov 2025"
.SH NAME
pdftotext \- Portable Document Format (PDF) to text converter
(version 4.06)
.SH SYNOPSIS
.B pdftotext
[options]
.RI [ PDF-file
.RI [ text-file ]]
.SH DESCRIPTION
.B Pdftotext
converts Portable Document Format (PDF) files to plain text.
.PP
Pdftotext reads the PDF file,
.IR PDF-file ,
and writes a text file,
.IR text-file .
If
.I text-file
is not specified, pdftotext converts
.I file.pdf
to
.IR file.txt .
If 
.I text-file
is \'-', the text is sent to stdout.
.SH CONFIGURATION FILE
Pdftotext reads a configuration file at startup.  It first tries to
find the user's private config file, ~/.xpdfrc.  If that doesn't
exist, it looks for a system-wide config file, typically /etc/xpdfrc
(but this location can be changed when pdftotext is built).  See the
.BR xpdfrc (5)
man page for details.
.SH OPTIONS
Many of the following options can be set with configuration file
commands.  These are listed in square brackets with the description of
the corresponding command line option.
.TP
.BI \-f " number"
Specifies the first page to convert.
.TP
.BI \-l " number"
Specifies the last page to convert.
.TP
.B \-layout
Maintain (as best as possible) the original physical layout of the
text.  The default is to \'undo' physical layout (columns,
hyphenation, etc.) and output the text in reading order.  If the
.B \-fixed
option is given, character spacing within each line will be determined
by the specified character pitch.
.TP
.B \-simple
Similar to
.BR \-layout ,
but optimized for simple one-column pages.  This mode will do a better
job of maintaining horizontal spacing, but it will only work properly
with a single column of text.
.TP
.B \-simple2
Similar to
.BR \-simple ,
but handles slightly rotated text (e.g., OCR output) better.  Only works
for pages with a single column of text.
.TP
.B \-table
Table mode is similar to physical layout mode, but optimized for
tabular data, with the goal of keeping rows and columns aligned (at
the expense of inserting extra whitespace).  If the
.B \-fixed
option is given, character spacing within each line will be determined
by the specified character pitch.
.TP
.B \-lineprinter
Line printer mode uses a strict fixed-character-pitch and -height
layout.  That is, the page is broken into a grid, and characters are
placed into that grid.  If the grid spacing is too small for the
actual characters, the result is extra whitespace.  If the grid
spacing is too large, the result is missing whitespace.  The grid
spacing can be specified using the
.B \-fixed
and
.B \-linespacing
options.
If one or both are not given on the command line, pdftotext will
attempt to compute appropriate value(s).
.TP
.B \-raw
Keep the text in content stream order.  Depending on how the PDF file
was generated, this may or may not be useful.
.TP
.BI \-fixed " number"
Specify the character pitch (character width), in points, for physical
layout, table, or line printer mode.  This is ignored in all other
modes.
.TP
.BI \-linespacing " number"
Specify the line spacing, in points, for line printer mode.  This is
ignored in all other modes.
.TP
.B \-clip
Text which is hidden because of clipping is removed before doing
layout, and then added back in.  This can be helpful for tables where
clipped (invisible) text would overlap the next column.
.TP
.B \-nodiag
Diagonal text, i.e., text that is not close to one of the 0, 90, 180,
or 270 degree axes, is discarded.  This is useful to skip watermarks
drawn on top of body text, etc.
.TP
.BI \-enc " encoding-name"
Sets the encoding to use for text output.  The
.I encoding\-name
must be defined with the unicodeMap command (see
.BR xpdfrc (5)).
The encoding name is case-sensitive.  This defaults to "Latin1" (which
is a built-in encoding).
.RB "[config file: " textEncoding ]
.TP
.BI \-eol " unix | dos | mac"
Sets the end-of-line convention to use for text output.
.RB "[config file: " textEOL ]
.TP
.B \-nopgbrk
Don't insert a page breaks (form feed character) at the end of each
page.
.RB "[config file: " textPageBreaks ]
.TP
.B \-bom
Insert a Unicode byte order marker (BOM) at the start of the text
output.
.TP
.BI \-marginl " number"
Specifies the left margin, in points.  Text in the left margin (i.e.,
within that many points of the left edge of the page) is discarded.
The default value is zero.
.TP
.BI \-marginr " number"
Specifies the right margin, in points.  Text in the right margin
(i.e., within that many points of the right edge of the page) is
discarded.  The default value is zero.
.TP
.BI \-margint " number"
Specifies the top margin, in points.  Text in the top margin (i.e.,
within that many points of the top edge of the page) is discarded.
The default value is zero.
.TP
.BI \-marginb " number"
Specifies the bottom margin, in points.  Text in the bottom margin
(i.e., within that many points of the bottom edge of the page) is
discarded.  The default value is zero.
.TP
.BI \-opw " password"
Specify the owner password for the PDF file.  Providing this will
bypass all security restrictions.
.TP
.BI \-upw " password"
Specify the user password for the PDF file.
.TP
.B \-verbose
Print a status message (to stdout) before processing each page.
.RB "[config file: " printStatusInfo ]
.TP
.B \-q
Don't print any messages or errors.
.RB "[config file: " errQuiet ]
.TP
.BI \-cfg " config-file"
Read
.I config-file
in place of ~/.xpdfrc or the system-wide config file.
.TP
.B \-listencodings
List all available text output encodings, then exit.
.TP
.B \-v
Print copyright and version information, then exit.
.TP
.B \-h
Print usage information, then exit.
.RB ( \-help
and
.B \-\-help
are equivalent.)
.SH BUGS
Some PDF files contain fonts whose encodings have been mangled beyond
recognition.  There is no way (short of OCR) to extract text from
these files.
.SH EXIT CODES
The Xpdf tools use the following exit codes:
.TP
0
No error.
.TP
1
Error opening a PDF file.
.TP
2
Error opening an output file.
.TP
3
Error related to PDF permissions.
.TP
98
Out of memory.
.TP
99
Other error.
.SH AUTHOR
The pdftotext software and documentation are copyright 1996-2025 Glyph
& Cog, LLC.
.SH "SEE ALSO"
.BR xpdf (1),
.BR pdftops (1),
.BR pdftohtml (1),
.BR pdfinfo (1),
.BR pdffonts (1),
.BR pdfdetach (1),
.BR pdftoppm (1),
.BR pdftopng (1),
.BR pdfimages (1),
.BR xpdfrc (5)
.br
.B http://www.xpdfreader.com/


ChangeLog
2025-10-16  Karl Berry  <karl@freefriends.org>

	* Build (status): avoid using egrep.
	Report from Jan Kasprzak,
	https://tug.org/pipermail/tex-k/2025-October/004217.html.

2025-09-07  Karl Berry  <karl@freefriends.org>

	* Makefile.am (install_target): new variable, install-strip by default.
	(world): use it.

2025-07-22  Karl Berry  <karl@freefriends.org>

	* reautoconf: mention that autoconf-only packages need to be
	explicitly added; would be better to read kpse-pkgs.m4.
	Also reautoconf tree for kpse-setup.m4 changes.

2025-06-26  Karl Berry  <karl@freefriends.org>

	* automake 1.18.1.

2025-04-30  Karl Berry  <karl@freefriends.org>

	* Build: incorporate original CFLAGS, CXXFLAGS, OBJCXXFLAGS values.
	https://tug.org/pipermail/tlbuild/2025q2/005771.html

2025-04-29  Karl Berry  <karl@freefriends.org>

	* Build: use TL_CFLAGS, TL_CXXFLAGS, etc., for TL_BUILD_ENV
	if -g is not given.

2025-03-08  Karl Berry  <karl@tug.org>

	* version.ac: 2026/dev.

2025-03-07  Karl Berry  <karl@tug.org>

	* TL'25 release.

2025-01-12  Karl Berry  <karl@tug.org>

	* Build: also support --debug-more for -g -Og -ggdb3,
	per Luigi, as described on https://tug.org/texlive/build.html.

2024-07-12  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake 1.17.

2024-04-29  Karl Berry  <karl@freefriends.org>

	* Makefile.am (distdir): override default dist generation,
	since we don't try to keep make dist working any more.
	https://tug.org/pipermail/tlbuild/2024q2/005547.html

2023-12-31  Karl Berry  <karl@freefriends.org>

	* reautoconf: add -f / --force option to pass to autoreconf.

2023-12-26  Karl Berry  <karl@freefriends.org>

	* reautoconf: don't bother with environment variables;
	because the autotools have many subprograms, setting PATH
	is what makes sense.

2023-12-23  Karl Berry  <karl@tug.org>

	* reautoconf with autoconf-2.72.

2023-12-21  Karl Berry  <karl@freefriends.org>

	* reautoconf (do_it): abort the run if any command fails.

2023-02-25  Karl Berry  <karl@tug.org>

	* Build: show full uname -a.

2023-02-09  Karl Berry  <karl@freefriends.org>

	* version.ac: 2023, no more /dev.

2022-03-21  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): 2022-03-21 for TL'22.

2022-01-18  Karl Berry  <karl@freefriends.org>

	* version.ac: 2022, no more /dev.

2021-10-04  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake-1.16.5.

2021-06-27  Karl Berry  <karl@freefriends.org>

	* reautoconf: pass -Wno-obsolete by default, since Autoconf 2.70
	and later obsoletes many common macros. See m4/README for more.
	Also standardize shell usage, etc., but no other functional changes.

2021-04-17  Karl Berry  <karl@tug.org>

	* version.ac: now 2022/dev since TL'21 is released.

2021-03-23  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-23 for TL'21.

2021-02-17  Karl Berry  <karl@tug.org>

	* version.ac: 2021, no more /dev.

2020-05-28  Karl Berry  <karl@freefriends.org>

	* Makefile.am (world): do not descend into $(texlinks_dir) if it
	doesn't exist. Report from Andreas Scherer,
	https://tug.org/pipermail/tlbuild/2020q1/004614.html.

2020-05-14  Karl Berry  <karl@freefriends.org>

	* reautoconf for removal of poppler support.

2020-04-10  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2020-03-27  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-27 for TL'20.

2020-03-06  Karl Berry  <karl@tug.org>

	* .travis.yml: attempt to reenable musl build on github.

2020-01-31  Karl Berry  <karl@tug.org>

	* Makefile.am (texlinks_dir): now texk/texlive/tl_support
	(instead of tl_scripts).

2020-01-16  Karl Berry  <karl@freefriends.org>

	* version.ac (tex_live_version): remove /dev, preparing for TL'20.

2019-06-25  Karl Berry  <karl@freefriends.org>

	* Build: don't call buildinfo from here, the output when the CC
	etc. envvars are not set is too misleading. Just write UNAME.

2019-06-24  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh (compiler_version): new shell function;
	 ignore useless first "Configured:" line for gcc->cc on Macs.

2019-06-23  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: output the compiler/other names
	that we pass from the buildinfo target.
	* Makefile.am (buildinfo): pass OBJCXX, not OBJC.
	* Build: more maintainable exports of TL_*.

2019-06-22  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: rename from build-info.sh;
	forget /etc/issue;
	allow option to omit config.status output;
	do not dump environment;
	run lsb_release -a. We'll see ...
	* Build: use buildinfo.sh instead of inline commands (mostly).
	* Makefile.am (buildinfo): make dependency of all-recursive
	so that it runs first; rename from build-info; remove from world.

2019-06-06  Karl Berry  <karl@tug.org>

	* Makefile.am (world): make build-info.
	(build-info): new target to invoke doc/build-info.sh.
	(build_info_target): new variable.
	* Build: export TL_* variables so build-info.sh can report them.
	* doc/build-info.sh: new script to report compiler versions,
	dump environment, config.status, etc.

2019-05-03  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2019-04-07  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): let's hope for 2019-04-07.

2018-10-02  Karl Berry  <karl@tug.org>

	* Build: TL_MAKE_VERBOSE=VERBOSE=1 by default,
	so that test-suite.log will be shown on failure.
	https://lists.gnu.org/archive/html/automake/2018-09/msg00015.html

2018-04-28  Karl Berry  <karl@tug.org>

	* version.ac: 2019/dev.

2018-04-10  Karl Berry  <karl@tug.org>

	* README.*: update from new tlbuild.

2017-11-15  Karl Berry  <karl@tug.org>

	* *.test: use #!/bin/sh -vx. Add 

2017-03-21  Karl Berry  <karl@freefriends.org>

	* everywhere: use EXTRA_DIST=TLpatches instead of <pkgname>-PATCHES.

2017-03-18  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2017/dev => 2017 for TL17.

2016-11-16  Karl Berry  <karl@tug.org>

	* Makefile.am (install-exec-hook): warn that the

libs/xpdf/xpdf-src/xpdf/pdftotext.cc
//========================================================================
//
// pdftotext.cc
//
// Copyright 1997-2013 Glyph & Cog, LLC
//
//========================================================================

#include <aconf.h>
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>
#ifdef DEBUG_FP_LINUX
#  include <fenv.h>
#  include <fpu_control.h>
#endif
#include "gmem.h"
#include "gmempp.h"
#include "parseargs.h"
#include "GString.h"
#include "GList.h"
#include "GlobalParams.h"
#include "Object.h"
#include "Stream.h"
#include "Array.h"
#include "Dict.h"
#include "XRef.h"
#include "Catalog.h"
#include "Page.h"
#include "PDFDoc.h"
#include "TextOutputDev.h"
#include "CharTypes.h"
#include "UnicodeMap.h"
#include "TextString.h"
#include "Error.h"
#include "config.h"

static int firstPage = 1;
static int lastPage = 0;
static GBool physLayout = gFalse;
static GBool simpleLayout = gFalse;
static GBool simple2Layout = gFalse;
static GBool tableLayout = gFalse;
static GBool linePrinter = gFalse;
static GBool rawOrder = gFalse;
static double fixedPitch = 0;
static double fixedLineSpacing = 0;
static GBool clipText = gFalse;
static GBool discardDiag = gFalse;
static char textEncName[128] = "";
static char textEOL[16] = "";
static GBool noPageBreaks = gFalse;
static GBool insertBOM = gFalse;
static double marginLeft = 0;
static double marginRight = 0;
static double marginTop = 0;
static double marginBottom = 0;
static char ownerPassword[33] = "\001";
static char userPassword[33] = "\001";
static GBool verbose = gFalse;
static GBool quiet = gFalse;
static char cfgFileName[256] = "";
static GBool listEncodings = gFalse;
static GBool printVersion = gFalse;
static GBool printHelp = gFalse;

static ArgDesc argDesc[] = {
  {"-f",       argInt,      &firstPage,     0,
   "first page to convert"},
  {"-l",       argInt,      &lastPage,      0,
   "last page to convert"},
  {"-layout",  argFlag,     &physLayout,    0,
   "maintain original physical layout"},
  {"-simple",  argFlag,     &simpleLayout,  0,
   "simple one-column page layout"},
  {"-simple2", argFlag,     &simple2Layout, 0,
   "simple one-column page layout, version 2"},
  {"-table",   argFlag,     &tableLayout,   0,
   "similar to -layout, but optimized for tables"},
  {"-lineprinter", argFlag, &linePrinter,   0,
   "use strict fixed-pitch/height layout"},
  {"-raw",     argFlag,     &rawOrder,      0,
   "keep strings in content stream order"},
  {"-fixed",   argFP,       &fixedPitch,    0,
   "assume fixed-pitch (or tabular) text"},
  {"-linespacing", argFP,   &fixedLineSpacing, 0,
   "fixed line spacing for LinePrinter mode"},
  {"-clip",    argFlag,     &clipText,      0,
   "separate clipped text"},
  {"-nodiag",  argFlag,     &discardDiag,   0,
   "discard diagonal text"},
  {"-enc",     argString,   textEncName,    sizeof(textEncName),
   "output text encoding name"},
  {"-eol",     argString,   textEOL,        sizeof(textEOL),
   "output end-of-line convention (unix, dos, or mac)"},
  {"-nopgbrk", argFlag,     &noPageBreaks,  0,
   "don't insert a page break at the end of each page"},
  {"-bom",     argFlag,     &insertBOM,     0,
   "insert a Unicode BOM at the start of the text file"},
  {"-marginl", argFP,       &marginLeft,    0,
   "left page margin"},
  {"-marginr", argFP,       &marginRight,   0,
   "right page margin"},
  {"-margint", argFP,       &marginTop,     0,
   "top page margin"},
  {"-marginb", argFP,       &marginBottom,  0,
   "bottom page margin"},
  {"-opw",     argString,   ownerPassword,  sizeof(ownerPassword),
   "owner password (for encrypted files)"},
  {"-upw",     argString,   userPassword,   sizeof(userPassword),
   "user password (for encrypted files)"},
  {"-verbose", argFlag,    &verbose,       0,
   "print per-page status information"},
  {"-q",       argFlag,     &quiet,         0,
   "don't print any messages or errors"},
  {"-cfg",     argString,   cfgFileName,    sizeof(cfgFileName),
   "configuration file to use in place of .xpdfrc"},
  {"-listencodings", argFlag, &listEncodings, 0,
   "list all available output text encodings"},
  {"-v",       argFlag,     &printVersion,  0,
   "print copyright and version info"},
  {"-h",       argFlag,     &printHelp,     0,
   "print usage information"},
  {"-help",    argFlag,     &printHelp,     0,
   "print usage information"},
  {"--help",   argFlag,     &printHelp,     0,
   "print usage information"},
  {"-?",       argFlag,     &printHelp,     0,
   "print usage information"},
  {NULL}
};

int main(int argc, char *argv[]) {
#if USE_EXCEPTIONS
  try {
#endif

  PDFDoc *doc;
  char *fileName;
  GString *textFileName;
  GString *ownerPW, *userPW;
  TextOutputControl textOutControl;
  TextOutputDev *textOut;
  UnicodeMap *uMap;
  GBool ok;
  char *p;
  int exitCode;

#ifdef DEBUG_FP_LINUX
  // enable exceptions on floating point div-by-zero
  feenableexcept(FE_DIVBYZERO);
  // force 64-bit rounding: this avoids changes in output when minor
  // code changes result in spills of x87 registers; it also avoids
  // differences in output with valgrind's 64-bit floating point
  // emulation (yes, this is a kludge; but it's pretty much
  // unavoidable given the x87 instruction set; see gcc bug 323 for
  // more info)
  fpu_control_t cw;
  _FPU_GETCW(cw);
  cw = (fpu_control_t)((cw & ~_FPU_EXTENDED) | _FPU_DOUBLE);
  _FPU_SETCW(cw);
#endif

  exitCode = 99;

  // parse args
  fixCommandLine(&argc, &argv);
  ok = parseArgs(argDesc, &argc, argv);
  if (ok && listEncodings) {
    // list available encodings
    if (cfgFileName[0] && !pathIsFile(cfgFileName)) {
      error(errConfig, -1, "Config file '{0:s}' doesn't exist or isn't a file",
	    cfgFileName);
    }
    globalParams = new GlobalParams(cfgFileName);
    GList *encs = globalParams->getAvailableTextEncodings();
    for (int i = 0; i < encs->getLength(); ++i) {
      printf("%s\n", ((GString *)encs->get(i))->getCString());
    }
    deleteGList(encs, GString);
    delete globalParams;
    goto err0;
  }
  if (printVersion) {
    printf("pdftotext version %s [www.xpdfreader.com]\n", xpdfVersion);
    printf("%s\n", xpdfCopyright);
    goto err0;
  }
  if (!ok || argc < 2 || argc > 3 || printHelp) {
    fprintf(stderr, "pdftotext version %s [www.xpdfreader.com]\n", xpdfVersion);
    fprintf(stderr, "%s\n", xpdfCopyright);
    printUsage("pdftotext", "<PDF-file> [<text-file>]", argDesc);
    goto err0;
  }
  fileName = argv[1];

  // read config file
  if (cfgFileName[0] && !pathIsFile(cfgFileName)) {
    error(errConfig, -1, "Config file '{0:s}' doesn't exist or isn't a file",
	  cfgFileName);
  }
  globalParams = new GlobalParams(cfgFileName);
  if (textEncName[0]) {
    globalParams->setTextEncoding(textEncName);
  }
  if (textEOL[0]) {
    if (!globalParams->setTextEOL(textEOL)) {
      fprintf(stderr, "Bad '-eol' value on command line\n");
    }
  }
  if (noPageBreaks) {
    globalParams->setTextPageBreaks(gFalse);
  }
  if (verbose) {
    globalParams->setPrintStatusInfo(verbose);
  }
  if (quiet) {
    globalParams->setErrQuiet(quiet);
  }

  // get mapping to output encoding
  if (!(uMap = globalParams->getTextEncoding())) {
    error(errConfig, -1, "Couldn't get text encoding");
    goto err1;
  }

  // open PDF file
  if (ownerPassword[0] != '\001') {
    ownerPW = new GString(ownerPassword);
  } else {
    ownerPW = NULL;
  }
  if (userPassword[0] != '\001') {
    userPW = new GString(userPassword);
  } else {
    userPW = NULL;
  }
  doc = new PDFDoc(fileName, ownerPW, userPW);
  if (userPW) {
    delete userPW;
  }
  if (ownerPW) {
    delete ownerPW;
  }
  if (!doc->isOk()) {
    exitCode = 1;
    goto err2;
  }

  // check for copy permission
  if (!doc->okToCopy()) {
    error(errNotAllowed, -1,
	  "Copying of text from this document is not allowed.");
    exitCode = 3;
    goto err2;
  }

  // construct text file name
  if (argc == 3) {
    textFileName = new GString(argv[2]);
  } else {
    p = fileName + strlen(fileName) - 4;
    if (strlen(fileName) > 4 && (!strcmp(p, ".pdf") || !strcmp(p, ".PDF"))) {
      textFileName = new GString(fileName, (int)strlen(fileName) - 4);
    } else {
      textFileName = new GString(fileName);
    }
    textFileName->append(".txt");
  }
  if (textFileName->cmp("-") == 0) {
    globalParams->setPrintStatusInfo(gFalse);
  }

  // get page range
  if (firstPage < 1) {
    firstPage = 1;
  }
  if (lastPage < 1 || lastPage > doc->getNumPages()) {
    lastPage = doc->getNumPages();
  }

  // write text file
  if (tableLayout) {
    textOutControl.mode = textOutTableLayout;
    textOutControl.fixedPitch = fixedPitch;
  } else if (physLayout) {
    textOutControl.mode = textOutPhysLayout;
    textOutControl.fixedPitch = fixedPitch;
  } else if (simpleLayout) {
    textOutControl.mode = textOutSimpleLayout;
  } else if (simple2Layout) {
    textOutControl.mode = textOutSimple2Layout;
  } else if (linePrinter) {
    textOutControl.mode = textOutLinePrinter;
    textOutControl.fixedPitch = fixedPitch;
    textOutControl.fixedLineSpacing = fixedLineSpacing;
  } else if (rawOrder) {
    textOutControl.mode = textOutRawOrder;
  } else {
    textOutControl.mode = textOutReadingOrder;
  }
  textOutControl.clipText = clipText;
  textOutControl.discardDiagonalText = discardDiag;
  textOutControl.insertBOM = insertBOM;
  textOutControl.marginLeft = marginLeft;
  textOutControl.marginRight = marginRight;
  textOutControl.marginTop = marginTop;
  textOutControl.marginBottom = marginBottom;
  textOut = new TextOutputDev(textFileName->getCString(), &textOutControl,
			      gFalse, gTrue);
  if (textOut->isOk()) {
    doc->displayPages(textOut, NULL, firstPage, lastPage, 72, 72, 0,
		      gFalse, gTrue, gFalse);
  } else {
    delete textOut;
    exitCode = 2;
    goto err3;
  }
  delete textOut;

  exitCode = 0;

  // clean up
 err3:
  delete textFileName;
 err2:
  delete doc;
  uMap->decRefCnt();
 err1:
  delete globalParams;
 err0:

  // check for memory leaks
  Object::memCheck(stderr);
  gMemReport(stderr);

  return exitCode;

#if USE_EXCEPTIONS
  } catch (GMemException e) {
    fprintf(stderr, "Out of memory\n");
    return 98;
  }
#endif
}
